### PR TITLE
Creation of explicit forwarding rules in the protect_payload.rs

### PR DESCRIPTION
**Pull request by francois141**
This commit creates a small framework similar to IPRules that allows to explicitly forwards registers from the payload to the firmware and in the opposite direction. Last but not least, it creates and test a forwarding rule for ecall from S mode.

### CODE REVIEW

File: src/policy/protect_payload.rs [hunk 1]
```diff
@@ -212,3 +216,87 @@ impl ProtectPayloadPolicy {
...
+            .allow_register_in(Register::X15)
+            .allow_register_in(Register::X16)
+            .allow_register_in(Register::X17)
+            .allow_register_out(Register::X10)
+            .allow_register_out(Register::X11);
+
+        rules
+    }
+
+    #[allow(unused)]
```
Comment (CharlyCst) [inline src/policy/protect_payload.rs:261]: It is used so no need for this annotation.

File: src/policy/protect_payload.rs [hunk 2]
```diff
@@ -212,3 +216,87 @@ impl ProtectPayloadPolicy {
...
+    fn allow_register_out(&mut self, reg: Register) -> &mut Self {
+        self.allow_out[reg as usize] = true;
+        self
+    }
+
+    #[allow(unused)]
+    fn block_register_out(&mut self, reg: Register) -> &mut Self {
+        self.allow_out[reg as usize] = false;
+        self
+    }
```
Comment (CharlyCst) [inline src/policy/protect_payload.rs:301]: In general, let's remove the `#[allow]` annotation and only keep the one we actually use.

The problem with those `#[allow]` annotation is that we always end-up forgetting to remove then, and then we end-up with a lot of dead-code that the compile doesn't warn about...

**Review by CharlyCst — COMMENTED** [review]
Looks cool, I mostly have some cleanup remarks :)

File: payload/test_protect_payload_payload/main.rs [hunk 3]
```diff
@@ -6,6 +6,76 @@
...
+impl Regs {
+    pub fn new() -> Self {
+        Regs { registers: [0; 32] }
+    }
+}
+
+macro_rules! save_registers {
+    ($regs:expr) => {
+            asm!(
+                "sd x0, 0*8({0})",
```
[the remark this answers is not in the document]
Comment (francois141) [reply payload/test_protect_payload_payload/main.rs:26]: True :)

**Pull request closed without merging**